src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/InstantiateArgumentExceptionsCorrectly.cs

```diff
@@ -125,26 +142,28 @@ public override void Initialize(AnalysisContext analysisContext)
         {
             bool matchesParameter = MatchesParameter(targetSymbol, creation, stringArgument);
             DiagnosticDescriptor? rule = null;
+            var dictBuilder = ImmutableDictionary.CreateBuilder<string, string?>();
 
             if (IsMessage(parameter) && matchesParameter)
             {
                 rule = RuleIncorrectMessage;
+                dictBuilder.Add(MessagePosition, parameter.Ordinal.ToString(CultureInfo.InvariantCulture));
             }
-            else if (IsParameterName(parameter) && !matchesParameter)
+            else if (HasParameters(targetSymbol) && IsParameterName(parameter) && !matchesParameter)
             {
                 // Allow argument exceptions in accessors to use the associated property symbol name.
-                if (MatchesAssociatedSymbol(targetSymbol, stringArgument))
+                if (!MatchesAssociatedSymbol(targetSymbol, stringArgument))
                 {
-                    return;
+                    rule = RuleIncorrectParameterName;
                 }
-
-                rule = RuleIncorrectParameterName;
             }
 
             if (rule != null)
             {
-                context.ReportDiagnostic(context.Operation.Syntax.CreateDiagnostic(rule, targetSymbol.Name, stringArgument, parameter.Name, creation.Type.Name));
+                return context.Operation.Syntax.CreateDiagnostic(rule, dictBuilder.ToImmutable(), targetSymbol.Name, stringArgument, parameter.Name, creation.Type.Name);
```

Comment:
Nit: Just a less verbose way to get the same functionality:
```suggestion
                return context.Operation.CreateDiagnostic(rule, dictBuilder.ToImmutable(), targetSymbol.Name, stringArgument, parameter.Name, creation.Type.Name);
```